### ISSUE TITLE
Rebus.Autofac can't resolve generics IFailed<T>, making SecondLevelTry to fail

### ISSUE THREAD

**Issue by pelin** [rebus-org/Rebus.Autofac#11]
When GetHandlerInvokers in ActivateHandlersStep for a failed message (FailedMessageWrapper), no invokers can be found, resulting in an empy list of handlers.

Register in Rebus.Autofac:
autofacBus.Subscribe<string>().Wait();
autofacBus.Subscribe<IFailed<string>>().Wait();
--> Do not work!

Register with BuildInActivator:
rebusActivator.Bus.Subscribe<string>();
rebusActivator.Bus.Subscribe<IFailed<string>>();
-- Working OK!

**Comment by mookid8000** [comment]
First off, the lines

    autofacBus.Subscribe<IFailed<string>>()

and

    rebusActivator.Bus.Subscribe<IFailed<string>>();

do not make any difference, because there's no events published to the topic for `IFailed<string>` (unless you explicitly went and did this, but I assume you didn't 😄 ).

When a message is in an endpoint's input queue, the message will be processed by that endpoint.

And then, if the message has failed too many times, and you have enabled 2nd level retries, then the message gets processed again – not as itself, but wrapped in an implementation of `IFailed<>`.

This means that your IoC container must be capable of resolving all `IHandleMessages<IFailed<YourMessage>>` implementations registered in your container.

My guess is that you have not registered your handler(s) in such a way that they're correctly resolved as implementations of `IHandleMessages<IFailed<YourMessage>>`.

Could you try and do this:

```csharp
var handlers = container.Resolve<IEnumerable<IHandleMessages<IFailed<YourMessage>>>>();
```
and then see what your Autofac container returns?

**Comment by pelin** [comment]
Thanks for your answer, and apologies for late response, but it's been vacationtime.

Still will say that this does not work. I tested the resolvement of the registration, and it do resolve. Check this "test code"


            // Rebus With AUTOFAC
            var builder = new ContainerBuilder();
            var rebusConfigurer = new RebusContainerConfigurer();
            builder.RegisterHandler<TestMessageRebusHandler>();
            builder.Register<ILogger>(c => new ConsoleLogger()).InstancePerDependency();
            var config = new RebusMessagingConfiguration("xyz...")
            {
                MaxRetries = 1,
                EnableSecondLevelRetries = true
            };
            var gatekeeperService = new GatekeeperServiceProxyFactory().Create();
            Action<OptionsConfigurer> gatekeeperOption = (OptionsConfigurer x) => x.ConfigureGatekeeperIncomingStep(gatekeeperService, logger);
            rebusConfigurer.RegisterRebusWithContainer(builder, config, null, logger, new List<Action<OptionsConfigurer>>() { gatekeeperOption }, null);
            var container = builder.Build();
            var autofacBus = container.Resolve<IBus>();
            autofacBus.Subscribe<string>().Wait();
            autofacBus.Subscribe<IFailed<string>>().Wait();
            var handlers = container.Resolve<IEnumerable<IHandleMessages<IFailed<string>>>>();
            // --> YES, IT RESOLVES!
            autofacBus.Publish("hi from Autofac").Wait();

            // Rebus With BuiltinHandlerActivator
            var rebusActivator = new BuiltinHandlerActivator();
            var bus = new ReBus();
            rebusActivator.Register(() => new TestMessageRebusHandler(logger, bus));
            var rebusMessagingConfiguration = new RebusMessagingConfiguration("xyz...")
            {
                AutomaticallyRenewLock = true,
                MaxRetries = 1,
                EnableSecondLevelRetries = true
            };
            var rebusBus = bus.Start(rebusActivator, rebusMessagingConfiguration);
            //Confa prenumerationer
            rebusActivator.Bus.Subscribe<string>();
            rebusActivator.Bus.Subscribe<IFailed<string>>();
            bus.Activate(rebusActivator);

            rebusBus.Publish("hi from BuiltInActivator").Wait();

The message handler class looks like this:

public class TestMessageRebusWithFailedHandler : IHandleMessages<string>, IHandleMessages<IFailed<string>>
    {
        private ILogger _logger;
        private IIntegrationBus _bus;
        public TestMessageRebusWithFailedHandler(ILogger logger, IIntegrationBus bus)
        {
            _logger = logger;
            _bus = bus;
        }

        public async Task Handle(string message)
        {
            _logger.LogDebug("String Message received!");

            throw new ApplicationException("test, should retry x times");
        }

        public async Task Handle(IFailed<string> message)
        {
            _logger.LogDebug("SLT:String  FailedMessage received!");
        }
    }


The BuiltInActivator works correctly and an IFailed-message is dispatched to the IFailed-handle-method, the Autofac-version does not.

**Comment by pelin** [comment]
Yes something was tricking me, and I 'm pretty sure I found what. This problem only happens in the beta-version (6.0.0-b01) of the Rebus.Autofac-package. Sorry I missed that important thing, that we ran the beta-version. I appreciate your time to investigate this, and you now have a headsup before releasing the beta-version of the rebus.autofac-package.. :-) We can degrade to the main-version so there is no hurry for us now. Thanks. //Peter
